### PR TITLE
Fixed zero insertion when concatenate sparse matrices with 0*I

### LINKED ISSUE #21563: 0I should not insert zeros when catted with sparse matrices
```
julia> [speye(3,3) 0I]
3×6 SparseMatrixCSC{Float64,Int64} with 6 stored entries:
  [1, 1]  =  1.0
  [2, 2]  =  1.0
  [3, 3]  =  1.0
  [1, 4]  =  0.0
  [2, 5]  =  0.0
  [3, 6]  =  0.0
```

Those stored zero values shouldn't be there.

### PR DESCRIPTION
**Pull request by iamnapo**
Fixes #21563

### CODE REVIEW

**Comment by tkelman** [comment]
LGTM, but is a behavior change so should merge post-branching

**Comment by iamnapo** [comment]
> merge post-branching

What do you mean?

**Comment by StefanKarpinski** [comment]
@tkelman means that we shouldn't merge this change until after the release-0.6 branch is created because this is a user-visible behavior change (albeit not likely to be a terribly breaking one).

**Comment by iamnapo** [comment]
Oh, cool.

File: test/sparse/sparse.jl [hunk 1]
```diff
@@ -47,12 +47,18 @@ end
 @testset "concatenation tests" begin
     @testset "horizontal concatenation" begin
         @test all([se33 se33] == sparse([1, 2, 3, 1, 2, 3], [1, 2, 3, 4, 5, 6], ones(6)))
+
+        sp33 = speye(3, 3)
```
Comment (Sacha0) [inline test/sparse/sparse.jl]: This `sp33` definition seems unnecessary given its absence in the immediately following test, and likewise two lines below? Or am I missing something? (Also, why the additional empty lines?)

Comment (iamnapo) [inline test/sparse/sparse.jl]: my bad, thanks!

File: test/sparse/sparse.jl [hunk 2]
```diff
@@ -45,14 +45,18 @@ do33 = ones(3)
 end
 
 @testset "concatenation tests" begin
+    sp33 = speye(3, 3)
+
     @testset "horizontal concatenation" begin
-        @test all([se33 se33] == sparse([1, 2, 3, 1, 2, 3], [1, 2, 3, 4, 5, 6], ones(6)))
+        @test all([se33 se33] == sparse([1, 2, 3, 1, 2, 3], [1, 2, 3, 4, 5, 6], ones(6)))
```
Comment (Sacha0) [inline test/sparse/sparse.jl]: Whitespace failure on this line?

**Review by stevengj — APPROVED** [review]

**Review by Sacha0 — APPROVED** [review]
Thanks @iamnapo! (Absent objections, requests for time, or someone beating me to it, I plan to merge this tomorrow morning. Best!)

**Review by StefanKarpinski — APPROVED** [review]

**Comment by Sacha0** [comment]
Thanks @iamnapo!

File: base/sparse/sparsematrix.jl [hunk 3]
```diff
@@ -1430,6 +1430,9 @@ speye_scaled(diag, m::Integer, n::Integer) = speye_scaled(typeof(diag), diag, m,
 
 function speye_scaled(T, diag, m::Integer, n::Integer)
     ((m < 0) || (n < 0)) && throw(ArgumentError("invalid array dimensions"))
+    if iszero(diag)
+        return SparseMatrixCSC(m, n, ones(Int, n+1), Vector{Int}(0), Vector{T}(0))
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: probably not very consequential, but I wonder how often this gets called in a way that makes the `Int(m), Int(n)` conversions below necessary

Comment (iamnapo) [inline base/sparse/sparsematrix.jl]: I think none. I even think the ones below are unnecessary, now that you mention it.

Comment (tkelman) [inline base/sparse/sparsematrix.jl]: maybe if you called it directly with a non Int sized integer?

Comment (iamnapo) [inline base/sparse/sparsematrix.jl]: Won't the constructor of `SparseMatrixCSC` take care of that?

Comment (stevengj) [inline base/sparse/sparsematrix.jl]: Yes, it looks like the `SparseMatrixCSC` constructor takes care of the `Int` conversions, so they could be removed here.

Comment (iamnapo) [inline base/sparse/sparsematrix.jl]: (I think is easier if someone changes it directly, instead of me opening another PR just for that.)

**Pull request merged**